jest/index.d.ts

```diff
@@ -46,6 +49,7 @@ declare namespace jest {
     /** Enables automatic mocking in the module loader. */
     function enableAutomock(): typeof jest;
     /** Creates a mock function. Optionally takes a mock implementation. */
+    function fn<T extends {}>(implementation: (...args: any[]) => T): Mock<T>;
     function fn<T>(implementation?: Function): Mock<T>;
```

Comment:
Yes, it's inferring only for constructor type when you're returning it from jest.fn(), i.e.
```js
const Mock = jest.fn(() => ({ prop: 1, method: jest.fn() });
const mock = new Mock();
mock.method();
mock.prop;
```
Second overload is for simple function mocks. ```extends {}``` allows this overload only be used if you're returning object from mock implementation